### ISSUE TITLE
On iOS12, app is stuck in the START state

### ISSUE THREAD

**Issue by shankari** [e-mission/e-mission-docs#735]
Encountered this while testing the NREL version of OpenPATH, but it seems like it could happen anywhere.

We tried to create the FSM before we had the right permissions (new authorization status = 0, always = 3), so the creation error failed as expected.

```
46,1654758641.0133,2022-06-09T00:10:41.013300-07:00,registered for visit notifications
47,1654758641.05402,2022-06-09T00:10:41.054020-07:00,initializing TripDiaryStateMachine with state = STATE_START
48,1654758641.07112,2022-06-09T00:10:41.071120-07:00,No precise location check needed for iOS < 14
49,1654758641.09456,2022-06-09T00:10:41.094560-07:00,Returning combination of always = 0 and precise 1
50,1654758641.25747,2022-06-09T00:10:41.257470-07:00,"In checker's didChangeAuthorizationStatus, new authorization status = 0, always = 3"
51,1654758641.27268,2022-06-09T00:10:41.272680-07:00,Calling TripDiarySettingsCheck from didChangeAuthorizationStatus to verify location service status and permission
52,1654758641.30666,2022-06-09T00:10:41.306660-07:00,No precise location check needed for iOS < 14
53,1654758641.3314302,2022-06-09T00:10:41.331430-07:00,Returning combination of always = 0 and precise 1
54,1654758641.3592598,2022-06-09T00:10:41.359260-07:00,Received platform-specific notification T_GEOFENCE_CREATION_ERROR
55,1654758641.37715,2022-06-09T00:10:41.377150-07:00,"In TripDiaryStateMachine, received transition T_GEOFENCE_CREATION_ERROR in state STATE_START"
56,1654758641.44459,2022-06-09T00:10:41.444590-07:00,Moved from STATE_START to STATE_START
```

Then, when we did get the correct permissions, the FSM did not get reinitialized

```
66,1654758657.30436,2022-06-09T00:10:57.304360-07:00,"in checkLocationSettingsAndPermissions, locationService is enabled, but the permission is 0"
67,1654758660.3803802,2022-06-09T00:11:00.380380-07:00,"In checker's didChangeAuthorizationStatus, new authorization status = 3, always = 3"
68,1654758660.40611,2022-06-09T00:11:00.406110-07:00,Calling TripDiarySettingsCheck from didChangeAuthorizationStatus to verify location service status and permission
69,1654758660.4525099,2022-06-09T00:11:00.452510-07:00,"In state STATE_START, Recieved location <+37.39093635,-122.08627154> +/- 1000.00m (speed -1.00 mps / course -1.00) @ 6/2/22, 5:30:49 PM Pacific Daylight Time"
```

we did get some weird prompt about incorrect app settings, which we ignored

```
76,1654758682.17226,2022-06-09T00:11:22.172260-07:00,Prompting for notification Incorrect app settings and text Click to view and fix app status
77,1654758682.18946,2022-06-09T00:11:22.189460-07:00,Notification triggered
78,1654758682.21035,2022-06-09T00:11:22.210350-07:00,"notification [object Object] is not an mode choice, returning..."
79,1654758685.67682,2022-06-09T00:11:25.676820-07:00,Ignoring notification Incorrect app settings and text Click to view and fix app status
```

The next TripDiaryStateMachine log was 

```
146,1654786613.5304198,2022-06-09T07:56:53.530420-07:00,"In TripDiaryStateMachine, received transition T_VISIT_ENDED in state STATE_START"
```

Attached logs:
[loggerDB.iphone4.notrips.withdate.log.gz](https://github.com/e-mission/e-mission-docs/files/8876638/loggerDB.iphone4.notrips.withdate.log.gz)
[loggerDB.iphone1.notrips.withdate.log.gz](https://github.com/e-mission/e-mission-docs/files/8876639/loggerDB.iphone1.notrips.withdate.log.gz)

**Comment by shankari** [comment]
Why did we not encounter this before? It seems like it would happen every time. Also why did it not happen on the iPhone7?

Couple of speculations:
- Did not happen on a large scale with CanBikeCO because they installed before the status screen
- Did not happen on the iPhone7 because the user paused after consenting and then continued with the login later?
- Did not happen with SULI intern test users because of ???

Deleted the app from one of the test phones and reinstalled.
Again, got a notification about incorrect app settings after accepting all permissions (remember seeing this last night as well)
Tried to handle instead of ignoring it this time, but it just popped up the status screen with all the entries green.
App again ended up in START state.

**Comment by shankari** [comment]
I vaguely remember encountering this error while testing the app screen on physical phones and running out of time to fix it then. I bet it is something to do with iOS12. And we haven't encountered it before since most of the other testers have late model iPhones running iOS15+.

Can't find a note to self recording this on the physical phone though.

A simple fix is to just drop support for iPhone12. But regardless, this is a generic platform issue, so let's move on to getting the NREL app in the stores for now.

**Comment by shankari** [comment]
Tested with the emulators and confirmed that the error occurs on iOS12 and not on iOS13. 
Maybe this is because in iOS13+, we open the settings screen, but in iOS12, we just show the popup.
At any rate, a simple fix is to drop support for iOS12.
A more equitable fix is to just always open the settings screen (if that works).

**Comment by asiripanich** [comment]
@shankari, I'm also seeing this with my fork of e-mission-phone master. I'm using iPhone 13 Pro running iOS 15.5. I built the app myself and uploaded it to test flight.

**Comment by shankari** [comment]
@asiripanich I have not been able to test on iOS 15.5 since I have to upgrade XCode which requires me to upgrade OSX. But I confirm that we do not see this behavior in iOS 15.2, or on any version since iOS 13+.

Are you sure you are seeing the same issue? Can you indicate what exactly you as you work through the permissions?

**Comment by shankari** [comment]
Actually, @sebastianbarry has an iOS 15.5 phone and he was able to install the NREL app without any problems. So this is not a generic problem with master.

**Comment by asiripanich** [comment]
Sure. I'm also experiencing the same issue on CanBikeCO.

**Comment by shankari** [comment]
Can you indicate what exactly you see as you work through the permissions? Because we have multiple 15.5 users on CanBikeCO who have not seen this problem.

**Comment by asiripanich** [comment]
Give me 10 minutes, let me video record the whole process.

**Comment by asiripanich** [comment]
![START STATE](https://user-images.githubusercontent.com/17020181/177259850-37d1d0c3-9d73-4b3f-926a-5c514a18dde6.gif)

**Comment by asiripanich** [comment]
And here is a "Check sensed" log from my own test flight app. I installed it 2 hours ago then went out to get lunch. It is still stuck at START. 

<img src="https://user-images.githubusercontent.com/17020181/177260296-d9128a97-31bb-495a-9ec7-84863cee8359.jpeg" 
     width="250" 
     height="500" />

**Comment by shankari** [comment]
when you type in the password, are you sure that it is the CanBikeCO app that is launched? I noticed that you had both CanBikeCO and FourStep installed at the same time, and typing in the password typically launches the first app installed (which is FourStep in your case).

**Comment by shankari** [comment]
You have uploaded a gif and not an mp4, so I can't see the URL that you are connecting to.

**Comment by shankari** [comment]
I should also fix this in general, but weighing the priority wrt other changes...

**Comment by shankari** [comment]
If you can reproduce this while installing CanBikeCO alone (without FourStep), can you email your CanBikeCO logs to me to see if it is the same problem? Because I don't see the incorrect app settings notification from 
https://github.com/e-mission/e-mission-docs/issues/735#issue-1267155505
in your video.

**Comment by asiripanich** [comment]
I can't find CanBikeCO in AU App Store.

**Comment by asiripanich** [comment]
> when you type in the password, are you sure that it is the CanBikeCO app that is launched? I noticed that you had both CanBikeCO and FourStep installed at the same time, and typing in the password typically launches the first app installed (which is FourStep in your case).

Yes, it was CanBikeCO.

**Comment by shankari** [comment]
> I can't find CanBikeCO in AU App Store.

Ah just added it.

**Comment by asiripanich** [comment]
> 

Tried this again using the test flight version of CanBikeCO. Still seeing the same issue.

**Comment by asiripanich** [comment]
> If you can reproduce this while installing CanBikeCO alone (without FourStep), can you email your CanBikeCO logs to me to see if it is the same problem? Because I don't see the incorrect app settings notification from
https://github.com/e-mission/e-mission-docs/issues/735#issue-1267155505
in your video.

Uploaded with name "Issue #735".

**Comment by shankari** [comment]
can you try with the version from the app store? it should be available now; I just added it.

**Comment by asiripanich** [comment]
> > I can't find CanBikeCO in AU App Store.
> 
> Ah just added it.

Might take awhile to appear on the AU app store.

**Comment by asiripanich** [comment]
> can you try with the version from the app store? it should be available now; I just added it.


If you can find its url please share it with me.

**Comment by asiripanich** [comment]
I can also try OpenPATH if you give me a token and a link to join a study.

**Comment by shankari** [comment]
QR code for the open access study:

![image](https://user-images.githubusercontent.com/2423263/177264936-f8ba01fb-da07-4848-9039-9fb3dd745f23.png)

You can use the autogenerated token when you get there.

> If you can find its url please share it with me.

https://apps.apple.com/us/app/id1526325846

**Comment by asiripanich** [comment]
> can you try with the version from the app store? it should be available now; I just added it.

Just uploaded another log with name "Issue #735 - App Store version".

**Comment by asiripanich** [comment]
@shankari can you confirm that after logging into the app, the expected state should be "WAITING_FOR_TRIP_START" not "START"?

**Comment by shankari** [comment]
@asiripanich yup, and that is what we see on iOS 13+ when we test

**Comment by asiripanich** [comment]
Well.. I'm seeing the same issue on OpenPATH.

**Comment by shankari** [comment]
Looking at your logs, it is in fact the same issue, except for the prompt about the incorrect app settings.

```
52,1657002380.3083901,2022-07-04T23:26:20.308390-07:00,registered for visit notifications
53,1657002380.30925,2022-07-04T23:26:20.309250-07:00,initializing TripDiaryStateMachine with state = STATE_START
54,1657002380.31031,2022-07-04T23:26:20.310310-07:00,Returning combination of always = 0 and precise 1
55,1657002380.31604,2022-07-04T23:26:20.316040-07:00,"In checker's didChangeAuthorizationStatus, new authorization status = 0, always = 3"
56,1657002380.3167,2022-07-04T23:26:20.316700-07:00,Calling TripDiarySettingsCheck from didChangeAuthorizationStatus to verify location service status and permission
57,1657002380.31923,2022-07-04T23:26:20.319230-07:00,Returning combination of always = 0 and precise 1
58,1657002380.32472,2022-07-04T23:26:20.324720-07:00,Received platform-specific notification T_GEOFENCE_CREATION_ERROR
59,1657002380.3253,2022-07-04T23:26:20.325300-07:00,"In TripDiaryStateMachine, received transition T_GEOFENCE_CREATION_ERROR in state STATE_START"
60,1657002380.32724,2022-07-04T23:26:20.327240-07:00,Moved from STATE_START to STATE_START
```

and

```
65,1657002387.66941,2022-07-04T23:26:27.669410-07:00,Returning combination of always = 0 and precise 1
66,1657002387.67705,2022-07-04T23:26:27.677050-07:00,"in checkLocationSettingsAndPermissions, locationService is enabled, but the permission is 0"
67,1657002388.4846199,2022-07-04T23:26:28.484620-07:00,Application went to the background
68,1657002389.70896,2022-07-04T23:26:29.708960-07:00,"Location tracking failed with error Error Domain=kCLErrorDomain Code=1 ""(null)"""
69,1657002389.73494,2022-07-04T23:26:29.734940-07:00,"In checker's didChangeAuthorizationStatus, new authorization status = 3, always = 3"
70,1657002389.7367,2022-07-04T23:26:29.736700-07:00,Calling TripDiarySettingsCheck from didChangeAuthorizationStatus to verify location service status and permission
71,1657002389.75498,2022-07-04T23:26:29.754980-07:00,"In state STATE_START, Recieved location ..."
72,1657002389.8241699,2022-07-04T23:26:29.824170-07:00,"In state STATE_START, Recieved location ..."
73,1657002391.29997,2022-07-04T23:26:31.299970-07:00,Application will enter the foreground
74,1657002391.3245,2022-07-04T23:26:31.324500-07:00,Returning combination of always = 1 and precise 1
75,1657002396.14972,2022-07-04T23:26:36.149720-07:00,"Motion activity available, checking auth status"
76,1657002396.15575,2022-07-04T23:26:36.155750-07:00,Auth status = 0
77,1657002396.15726,2022-07-04T23:26:36.157260-07:00,"Activity status not determined, initializing to get regular prompt"
78,1657002397.09724,2022-07-04T23:26:37.097240-07:00,activity recognition works fine
79,1657002401.17044,2022-07-04T23:26:41.170440-07:00,Received callback from user notification settings
80,1657002401.98144,2022-07-04T23:26:41.981440-07:00,"changing consent from null -> {""protocol_id"":""2014-04-6267"",""approval_date"":""2016-07-14""}"
81,1657002401.99527,2022-07-04T23:26:41.995270-07:00,"required consent version = {""protocol_id"":""2014-04-6267"",""approval_date"":""2016-07-14""}"
84,1657002402.0052302,2022-07-04T23:26:42.005230-07:00,"Initialized remote push notification handler <BEMRemotePushNotificationHandler: 0x281cac820>, finished registering for notifications "
```

**Comment by shankari** [comment]
Ok so this is super weird. I clearly documented that I tested this on iOS13+ and it worked ("confirmed that the error occurs on iOS12 and not on iOS13.")
https://github.com/e-mission/e-mission-docs/issues/735#issuecomment-1152094732
but when I try to re-test is now, it doesn't work.

And I apparently didn't save the logs from when I reproduced. And I still don't understand how it worked for @sebastianbarry - I am definitely seeing data coming in from him.

````
>>> edb.get_timeseries_db().find({"metadata.key": "background/filtered_location"}).count()
50023
````

It is midnight here and the fix is not likely to be super simple. And I want to figure out how this ever worked (with @sebastianbarry) to see if that gives us a clue about how we should fix it.

This is definitely my highest priority fix tomorrow morning...

**Comment by asiripanich** [comment]
> Ok so this is super weird. I clearly documented that I tested this on iOS13+ and it worked ("confirmed that the error occurs on iOS12 and not on iOS13.") [#735 (comment)](https://github.com/e-mission/e-mission-docs/issues/735#issuecomment-1152094732) but when I try to re-test is now, it doesn't work.
> 
> And I apparently didn't save the logs from when I reproduced. And I still don't understand how it worked for @sebastianbarry - I am definitely seeing data coming in from him.
> 
> ```
> >>> edb.get_timeseries_db().find({"metadata.key": "background/filtered_location"}).count()
> 50023
> ```
> 
> It is midnight here and the fix is not likely to be super simple. And I want to figure out how this ever worked (with @sebastianbarry) to see if that gives us a clue about how we should fix it.
> 
> This is definitely my highest priority fix tomorrow morning...

If it helps, the version of Fourstep (forked in Sep 2021) that I have on App store (not the one you see in my Test Flight screen) doesn't have this issue.

**Comment by shankari** [comment]
that version would not have the status screen, right? If this is a problem, it would be related to the status screen.
We have had multiple people install the app after the status screen changes, though, including all the summer interns, and none of them seem to have had this issue.

**Comment by asiripanich** [comment]
Yep. No status screen.

**Comment by shankari** [comment]
Just to briefly think through a potential fix since it looks like this is fairly reproducible at least for @asiripanich: the real problem is that we are trying to initialize the FSM before we even show the user the permission screen **and** we don't re-initialize the FSM once the user gives the permission.

Digging deeper:
- we initialize the FSM in `+ (TripDiaryStateMachine*) instance {` without checking for consent
- we check for consent in the plugin initializer

```
    BOOL isConsented = [ConfigManager isConsented:reqConsent];
    if (isConsented) {
        [self initWithConsent];
    } else {
```

- however, with the permission screen, we also introduce a "back door" initialization at

```
+(BOOL)checkLocationPermissions {
    BOOL alwaysPerm = [CLLocationManager authorizationStatus] == kCLAuthorizationStatusAuthorizedAlways;
    BOOL precisePerm = TRUE;
            CLLocationManager* currLocMgr = [TripDiaryStateMachine instance].locMgr;
            if (@available(iOS 14.0, *)) {
                CLAccuracyAuthorization preciseOrNot = [currLocMgr accuracyAuthorization];
        precisePerm = preciseOrNot == CLAccuracyAuthorizationFullAccuracy;
            } else {
                [LocalNotificationManager addNotification:[NSString stringWithFormat:@"No precise location check needed for iOS < 14"]];
            }
    [LocalNotificationManager addNotification:[NSString stringWithFormat:@"Returning combination of always = %@ and precise %@", @(alwaysPerm), @(precisePerm)]];
    return alwaysPerm && precisePerm;
}
```

So one fix would be to restructure that code to avoid the initialization, for example by checking for the precise authorization only when the always perm is set.

**Comment by shankari** [comment]
After making that change, we initialize the FSM while prompting for permission. We need this because if we don't first call `startUpdatingLocation`, location won't even show up
https://github.com/e-mission/e-mission-data-collection/commit/392c34e57c0d0b42d51bfd3f41ea9e68cf64de0e

```
-(void)promptForPermission:(CLLocationManager*)locMgr {
    if (IsAtLeastiOSVersion(@"13.0")) {
        NSLog(@"iOS 13+ detected, launching UI settings to easily enable always");
        // we want to leave the registration in the prompt for permission, since we don't want to register callbacks when we open the app settings for other reasons
        [[TripDiaryStateMachine delegate] registerForegroundDelegate:self];
        [[TripDiaryStateMachine instance].locMgr startUpdatingLocation];
        [self openAppSettings];
```

So that is not fixable. so we want to ensure that we can re-initalize the FSM once we do get the permissions, in logs like this:


```
2022-07-05 01:06:10.136968-0700 CanBikeCO[12962:7655381] Calling TripDiarySettingsCheck from didChangeAuthorizationStatus to verify location service status and permission
2022-07-05 01:06:10.137145-0700 CanBikeCO[12962:7655381] DEBUG: Calling TripDiarySettingsCheck from didChangeAuthorizationStatus to verify location service status and permission

2022-07-05 01:07:05.508332-0700 CanBikeCO[12962:7655381] In checker's didChangeAuthorizationStatus, new authorization status = 3, always = 3
2022-07-05 01:07:05.508600-0700 CanBikeCO[12962:7655381] DEBUG: In checker's didChangeAuthorizationStatus, new authorization status = 3, always = 3
2022-07-05 01:07:05.558565-0700 CanBikeCO[12962:7655381] Calling TripDiarySettingsCheck from didChangeAuthorizationStatus to verify location service status and permission
2022-07-05 01:07:05.558760-0700 CanBikeCO[12962:7655381] DEBUG: Calling TripDiarySettingsCheck from didChangeAuthorizationStatus to verify location service status and permission


2022-07-05 01:07:28.230051-0700 CanBikeCO[12962:7655381] Returning combination of always = 1 and precise 1
2022-07-05 01:07:28.230194-0700 CanBikeCO[12962:7655381] DEBUG: Returning combination of always = 1 and precise 1
```

I note also that there is some pending code around the status screen where we need to `Figure out how to fix recursion when [TripDiaryStateMachine instance] is called`

https://github.com/e-mission/e-mission-data-collection/blob/master/src/ios/Location/TripDiaryStateMachine.m#L107

**Comment by shankari** [comment]
Those logs are from

```
- (void)locationManager:(CLLocationManager *)manager
            didChangeAuthorizationStatus:(CLAuthorizationStatus)status {
    [LocalNotificationManager addNotification:[NSString stringWithFormat:@"In checker's didChangeAuthorizationStatus, new authorization status = %d, always = %d", status, kCLAuthorizationStatusAuthorizedAlways]];

    [LocalNotificationManager addNotification:[NSString stringWithFormat:@"Calling TripDiarySettingsCheck from didChangeAuthorizationStatus to verify location service status and permission"]];
    if (foregroundDelegateList.count > 0) {
        for (id currDelegate in foregroundDelegateList) {
            [currDelegate didChangeAuthorizationStatus:(CLAuthorizationStatus)status];
        }
        [foregroundDelegateList removeAllObjects];
    } else {
        [SensorControlBackgroundChecker checkAppState];
    }
}
```

However, in the foreground delegate callback, we have

```
- (void) didChangeAuthorizationStatus:(CLAuthorizationStatus)status
{
    [[TripDiaryStateMachine instance].locMgr stopUpdatingLocation];
    NSString* callbackId = [command callbackId];
    @try {
        if (status == kCLAuthorizationStatusAuthorizedAlways) {
            CDVPluginResult* result = [CDVPluginResult
                                       resultWithStatus:CDVCommandStatus_OK];
            [commandDelegate sendPluginResult:result callbackId:callbackId];
        } else {
            NSString* msg = NSLocalizedStringFromTable(@"location_permission_off_app_open",         @"DCLocalizable", nil);
            CDVPluginResult* result = [CDVPluginResult
                                       resultWithStatus:CDVCommandStatus_ERROR
                                       messageAsString:msg];
            [commandDelegate sendPluginResult:result callbackId:callbackId];
        }
    }
    @catch (NSException *exception) {
        NSString* msg = [NSString stringWithFormat: @"While handling auth callback, error %@", exception];
        CDVPluginResult* result = [CDVPluginResult
                                   resultWithStatus:CDVCommandStatus_ERROR
                                   messageAsString:msg];
        [commandDelegate sendPluginResult:result callbackId:callbackId];
    }

}
```

Maybe we can add a call to restart FSM if start state.

**Comment by shankari** [comment]
Just realized that I had a copy of @sebastianbarry's logs to investigate https://github.com/e-mission/e-mission-docs/issues/640

Let's see if that gives us any clues about how this worked for others.

I'm not sure I have all the logs, because the intro was already done when the logs start

```
10,1653508620.45545,2022-05-25T12:57:00.455450-07:00,"uc_stored_val = {""intro_done"":""2022-05-23T10:56:56-06:00""} ls_stored_val = {""intro_done"":""2022-05-23T10:56:56-06:00""}"
```

but the app was still in the `START` state

```
146,1653524632.01288,2022-05-25T17:23:52.012880-07:00,"In state STATE_START, Recieved location ..."
```

He logged in again 

```
86,1653508844.24329,2022-05-25T13:00:44.243290-07:00,Returning combination of always = 1 and precise 1
87,1653508844.25362,2022-05-25T13:00:44.253620-07:00,Returning combination of always = 1 and precise 1
90,1653508844.26324,2022-05-25T13:00:44.263240-07:00,postToHost called with url = /profile/create -- https://stage.canbikeco.org
```

```
205,1653524835.70696,2022-05-25T17:27:15.706960-07:00,Received platform-specific notification T_VISIT_ENDED
206,1653524835.7085,2022-05-25T17:27:15.708500-07:00,lastLocArray.length = 0 while generating trip start event
207,1653524835.71015,2022-05-25T17:27:15.710150-07:00,Broadcasting generic transition trip_started and generating notification
208,1653524835.71209,2022-05-25T17:27:15.712090-07:00,Generating all notifications for generic trip_started
209,1653524835.71405,2022-05-25T17:27:15.714050-07:00,"no configurations found for event trip_started, skipping notification"
210,1653524835.71584,2022-05-25T17:27:15.715840-07:00,"In TripDiaryStateMachine, received transition T_VISIT_ENDED in state STATE_START"
211,1653525152.3319302,2022-05-25T17:32:32.331930-07:00,"Received visit notification = <+39.67196639,-104.96206512> +/- 21.00m (2022-05-26 00:26:53 +0000 to -) CLPlaceInference, ref, ..."
```

and then, a few days later, he went through and changed the authorization again

```
335,1653578367.20367,2022-05-26T08:19:27.203670-07:00,"In checker's didChangeAuthorizationStatus, new authorization status = 3, always = 3"
336,1653578367.20469,2022-05-26T08:19:27.204690-07:00,Calling TripDiarySettingsCheck from didChangeAuthorizationStatus to verify location service status and permission
337,1653578367.2070901,2022-05-26T08:19:27.207090-07:00,Returning combination of always = 1 and precise 1
338,1653578367.2121801,2022-05-26T08:19:27.212180-07:00,"All settings valid, nothing to prompt"
339,1653578367.21354,2022-05-26T08:19:27.213540-07:00,"Still in start state, sending initialize..."
340,1653578367.21491,2022-05-26T08:19:27.214910-07:00,"In TripDiaryStateMachine, received transition T_INITIALIZE in state STATE_START"
341,1653578367.21858,2022-05-26T08:19:27.218580-07:00,registered for visit notifications
342,1653578367.2193902,2022-05-26T08:19:27.219390-07:00,In createGeofenceHere
```

and this time, these two messages caused the FSM to restart

```
338,1653578367.2121801,2022-05-26T08:19:27.212180-07:00,"All settings valid, nothing to prompt"
339,1653578367.21354,2022-05-26T08:19:27.213540-07:00,"Still in start state, sending initialize..."
```

Which happens when `[SensorControlBackgroundChecker checkAppState]` is called.

```
+(void)checkAppState
{
    if (allChecksPass) {
        [LocalNotificationManager addNotification:[NSString stringWithFormat:@"All settings valid, nothing to prompt"]];
        [self restartFSMIfStartState];
    }
}
```

and `[SensorControlBackgroundChecker checkAppState]` is called from the remote push code

```
+ (void) performPeriodicActivity
{
    [SensorControlBackgroundChecker checkAppState];
    [SensorControlBackgroundChecker restartFSMIfStartState];
}
```

and also from the auth change callback, if it was invoked from the background

```
- (void)locationManager:(CLLocationManager *)manager
            didChangeAuthorizationStatus:(CLAuthorizationStatus)status {
    [LocalNotificationManager addNotification:[NSString stringWithFormat:@"In checker's didChangeAuthorizationStatus, new authorization status = %d, always = %d", status, kCLAuthorizationStatusAuthorizedAlways]];

    [LocalNotificationManager addNotification:[NSString stringWithFormat:@"Calling TripDiarySettingsCheck from didChangeAuthorizationStatus to verify location service status and permission"]];
    if (foregroundDelegateList.count > 0) {
        for (id currDelegate in foregroundDelegateList) {
            [currDelegate didChangeAuthorizationStatus:(CLAuthorizationStatus)status];
        }
        [foregroundDelegateList removeAllObjects];
    } else {
        [SensorControlBackgroundChecker checkAppState];
    }
}
```

In this case, it seems to have been the latter and not the former, because of 

```
335,1653578367.20367,2022-05-26T08:19:27.203670-07:00,"In checker's didChangeAuthorizationStatus, new authorization status = 3, always = 3"
```

which seems to happen multiple times for this user

```
$ grep "didChangeAuthorizationStatus.*new authorization" /tmp/2022-06-23_testing_sebastian_1_America_Denver.loggerDB.withdate.log | wc -l
      84

$ grep "didChangeAuthorizationStatus.*new authorization" /tmp/2022-06-23_testing_sebastian_1_America_Denver.loggerDB.withdate.log | head -n 2
335,1653578367.20367,2022-05-26T08:19:27.203670-07:00,"In checker's didChangeAuthorizationStatus, new authorization status = 3, always = 3"
954,1653590732.4681098,2022-05-26T11:45:32.468110-07:00,"In checker's didChangeAuthorizationStatus, new authorization status = 3, always = 3"

grep "didChangeAuthorizationStatus.*new authorization" /tmp/2022-06-23_testing_sebastian_1_America_Denver.loggerDB.withdate.log | tail -n 2
194242,1656003294.3666599,2022-06-23T09:54:54.366660-07:00,"In checker's didChangeAuthorizationStatus, new authorization status = 3, always = 3"
195010,1656007549.89205,2022-06-23T11:05:49.892050-07:00,"In checker's didChangeAuthorizationStatus, new authorization status = 3, always = 3"
```

He also gets push notifications after the FSM is initialized correctly (not sure why not earlier)

```
$ grep "Found silent push notification, for platform ios" /tmp/2022-06-23_testing_sebastian_1_America_Denver.loggerDB.withdate.log | head -n 2
9010,1653797967.51459,2022-05-28T21:19:27.514590-07:00,"Found silent push notification, for platform ios"
10208,1653804709.7099,2022-05-28T23:11:49.709900-07:00,"Found silent push notification, for platform ios"

$ grep "Found silent push notification, for platform ios" /tmp/2022-06-23_testing_sebastian_1_America_Denver.loggerDB.withdate.log | tail -n 2
193010,1655967657.3888,2022-06-23T00:00:57.388800-07:00,"Found silent push notification, for platform ios"
193455,1656000541.49562,2022-06-23T09:09:01.495620-07:00,"Found silent push notification, for platform ios"
```

**Comment by shankari** [comment]
Got logs from @sebastianbarry's install of the NREL OpenPATH app. Again, the push notification saved the day.
I will still work on fixing this ASAP but will also try to batch up some changes before pushing it out instead of treating it as a fire drill :)

In state START

```
54,1656627997.274,2022-06-30T15:26:37.274000-07:00,"In checker's didChangeAuthorizationStatus, new authorization status = 0, always = 3"
55,1656627997.27509,2022-06-30T15:26:37.275090-07:00,Calling TripDiarySettingsCheck from didChangeAuthorizationStatus to verify location service status and permission
56,1656627997.27743,2022-06-30T15:26:37.277430-07:00,Returning combination of always = 0 and precise 1
57,1656627997.28018,2022-06-30T15:26:37.280180-07:00,"In TripDiaryStateMachine, received transition T_GEOFENCE_CREATION_ERROR in state STATE_START"
58,1656627997.28446,2022-06-30T15:26:37.284460-07:00,Moved from STATE_START to STATE_START
```

Still in state START

```
73,1656628007.56784,2022-06-30T15:26:47.567840-07:00,"In checker's didChangeAuthorizationStatus, new authorization status = 3, always = 3"
74,1656628007.66146,2022-06-30T15:26:47.661460-07:00,Calling TripDiarySettingsCheck from didChangeAuthorizationStatus to verify location service status and permission
75,1656628007.66921,2022-06-30T15:26:47.669210-07:00,"In state STATE_START, Recieved location..."
76,1656628007.7838,2022-06-30T15:26:47.783800-07:00,"In state STATE_START, Recieved location..."
```

still in state START

```
182,1656631603.3465,2022-06-30T16:26:43.346500-07:00,"In state STATE_START, Recieved location ..."
183,1656631603.8859699,2022-06-30T16:26:43.885970-07:00,"Received visit notification = ..."
184,1656631603.89398,2022-06-30T16:26:43.893980-07:00,"reqConsent = 2016-07-14, currConfig.approval_date = 2016-07-14"
185,1656631603.89675,2022-06-30T16:26:43.896750-07:00,isConsented = YES
186,1656631603.8995,2022-06-30T16:26:43.899500-07:00,"departure date is 2022-06-30 23:25:55 +0000, isDistantDate? 0, after distantDate? -1 "
187,1656631603.9017599,2022-06-30T16:26:43.901760-07:00,"In TripDiaryStateMachine, received transition T_VISIT_ENDED in state STATE_START"
```

Got push notification, which fixed everything

```
190,1656631604.75505,2022-06-30T16:26:44.755050-07:00,"Found silent push notification, for platform ios"
191,1656631604.7572699,2022-06-30T16:26:44.757270-07:00,"Platform is ios, calling handleSilentPush on DataCollection"
197,1656631604.77114,2022-06-30T16:26:44.771140-07:00,"In TripDiaryStateMachine, received transition T_RECEIVED_SILENT_PUSH in state STATE_START"
198,1656631604.77575,2022-06-30T16:26:44.775750-07:00,"In TripDiaryStateMachine, received transition T_NOP in state STATE_START"
199,1656631604.78172,2022-06-30T16:26:44.781720-07:00,Returning combination of always = 1 and precise 1
200,1656631604.79737,2022-06-30T16:26:44.797370-07:00,"All settings valid, nothing to prompt"
201,1656631604.7991,2022-06-30T16:26:44.799100-07:00,"Still in start state, sending initialize..."
202,1656631604.8008,2022-06-30T16:26:44.800800-07:00,"In TripDiaryStateMachine, received transition T_INITIALIZE in state STATE_START"
....
226,1656631605.3713698,2022-06-30T16:26:45.371370-07:00,Current state of region STATIONARY_GEOFENCE_LOCATION is 1 (inside)
227,1656631605.38237,2022-06-30T16:26:45.382370-07:00,Received INSIDE geofence state when currState = STATE_START
228,1656631605.3873198,2022-06-30T16:26:45.387320-07:00,"In TripDiaryStateMachine, received transition T_INIT_COMPLETE in state STATE_START"
229,1656631605.4072301,2022-06-30T16:26:45.407230-07:00,Moved from STATE_START to STATE_WAITING_FOR_TRIP_START
```

**Comment by asiripanich** [comment]
The screenshot is from my CanBikeCO, waited for 8 hours but no transition state change.

![image](https://user-images.githubusercontent.com/17020181/177429861-a4e7496e-0518-4bce-908d-392cdd24e239.png)

**Comment by asiripanich** [comment]
Uploaded the log file.

**Comment by shankari** [comment]
There are no push notifications received

```
$ grep silent /tmp/2022-07-05_735_Australia_Sydney.loggerDB.withdate.log | wc -l
       0

$ grep silent /tmp/sebastian_nrel_openpath.loggerDB.withdate.log | wc -l
     326
```

Not quite sure why you are not getting any push notifications in CanBikeCO. Are you connected to the network, for example?

**Comment by asiripanich** [comment]
Yes, I was connected to the network the whole time.

**Comment by shankari** [comment]
I tried sending you a visible push and it failed 

```
DEBUG:root:firebase push result for ios: success 0 failure 1 results [{'error': 'NotRegistered'}]
```

So presumably the silent push also failed.

**Comment by shankari** [comment]
The token in the logs

```
130,1657033962.8874,2022-07-05T08:12:42.887400-07:00,"Token = {""token"":""ebVkt1qfVECAi52u-B2hxN:...",""type"":""FCM""}"
134,1657033962.91117,2022-07-05T08:12:42.911170-07:00,"Finished saving token = ""ebVkt1qfVECAi52u-B2hxN:...""
```

doesn't match the token on the server

```
DEBUG:root:adding token chTHO6y-lko2tKLWBqK0GT:... to list for platform ios
```

**Comment by shankari** [comment]
@asiripanich which logs am I supposed to be looking at?

The logs that I see are

```
-rw-r--r-- 1 root   root       32768 Jul  5 23:05 2022-07-05_735_Australia_Sydney.loggerDB
-rw-r--r-- 1 root   root       28672 Jul  5 06:39 2022-07-05_Issue_735_-_App_Store_version_Australia_Sydney.loggerDB
-rw-r--r-- 1 root   root       28672 Jul  5 06:30 2022-07-05_Issue_735_Australia_Sydney.loggerDB
```

However, the last `/profile/update`, which set the token to `chTHO6y-lko2tKLWBqK0GT:` is at

```
2022-07-06 04:30:43,488:DEBUG:140091271923456:START POST /profile/update
2022-07-06 04:30:43,488:DEBUG:140091271923456:Called updateUserProfile
2022-07-06 04:30:43,488:DEBUG:140091271923456:methodName = token_list, returning <class 'emission.net.auth.token_list.TokenListMethod'>
2022-07-06 04:30:43,489:DEBUG:140091271923456:Using the TokenListMethod to verify id token ...
2022-07-06 04:30:43,489:DEBUG:140091271923456:Found match for token .... of length 17
2022-07-06 04:30:43,490:DEBUG:140091271923456:retUUID = 6871ab4c-...
2022-07-06 04:30:43,495:DEBUG:140091271923456:Updating user 6871ab4c-... with fields {'curr_platform': 'ios', 'curr_sync_interval': 3600, 'device_token': 'chTHO6y-lko2tKLWBqK0GT:...'}
2022-07-06 04:30:43,497:DEBUG:140091271923456:Successfully updated profile for user 6871ab4c-dc34-4257-8e90-29f84f2204f1
2022-07-06 04:30:43,498:DEBUG:140091271923456:END POST /profile/update 6871ab4c-...0.009972333908081055
```

So it doesn't look like we have the matching logs?

**Comment by asiripanich** [comment]
> I tried sending you a visible push and it failed
> 
> ```
> DEBUG:root:firebase push result for ios: success 0 failure 1 results [{'error': 'NotRegistered'}]
> ```
> 
> So presumably the silent push also failed.

I got the same error when I tried this on my own server as well.

**Comment by asiripanich** [comment]
> -rw-r--r-- 1 root   root       32768 Jul  5 23:05 2022-07-05_735_Australia_Sydney.loggerDB

This is from this morning after you asked me to keep the app on for a few hours.

**Comment by shankari** [comment]
ok, but none of them seems to overlap the time the token was updated, around `2022-07-06 04:30:43`

**Comment by shankari** [comment]
Moved the discussion about the push notifications to a new issue https://github.com/e-mission/e-mission-docs/issues/749

**Comment by shankari** [comment]
Continuing the fix from https://github.com/e-mission/e-mission-docs/issues/735#issuecomment-1174765992

> Maybe we can add a call to restart FSM if start state if the new authorization is "ALWAYS".

**Comment by shankari** [comment]
This doesn't work because when we click "fix" to fix the location permissions, we get a callback for the `didChangeAuthorizationStatus` with new authorization status = 0 (unauthorized).

```
2022-07-05 22:44:05.661499-0700 CanBikeCO[74178:8688001] iOS 13+ detected, launching UI settings to easily enable always
2022-07-05 22:44:05.663944-0700 CanBikeCO[74178:8688001] THREAD WARNING: ['DataCollection'] took '24.260986' ms. Plugin should use a background thread.
2022-07-05 22:44:05.664328-0700 CanBikeCO[74178:8688001] In checker's didChangeAuthorizationStatus, new authorization status = 0, always = 3
2022-07-05 22:44:05.664417-0700 CanBikeCO[74178:8688001] DEBUG: In checker's didChangeAuthorizationStatus, new authorization status = 0, always = 3
2022-07-05 22:44:05.665844-0700 CanBikeCO[74178:8688001] 1 foreground delegates found, calling didChangeAuthorizationStatus to return the new value 0
2022-07-05 22:44:05.665938-0700 CanBikeCO[74178:8688001] DEBUG: 1 foreground delegates found, calling didChangeAuthorizationStatus to return the new value 0
```

We pass this on to the delegate which returns the error

```
2022-07-05 22:44:05.667160-0700 CanBikeCO[74178:8688001] In delegate didChangeAuthorizationStatus, newStatus = 0
2022-07-05 22:44:05.667263-0700 CanBikeCO[74178:8688001] DEBUG: In delegate didChangeAuthorizationStatus, newStatus = 0
2022-07-05 22:44:05.668670-0700 CanBikeCO[74178:8688001] status 0 != always 3, returning error
2022-07-05 22:44:05.668744-0700 CanBikeCO[74178:8688001] DEBUG: status 0 != always 3, returning error
```

We then remove all foreground delegates.

```
2022-07-05 22:44:05.670079-0700 CanBikeCO[74178:8688001] Notified all foreground delegates, removing all of them
2022-07-05 22:44:05.670175-0700 CanBikeCO[74178:8688001] DEBUG: Notified all foreground delegates, removing all of them
```

So when we do get the always permission,

```
2022-07-05 22:47:07.048200-0700 CanBikeCO[74178:8688001] In checker's didChangeAuthorizationStatus, new authorization status = 3, always = 3
2022-07-05 22:47:07.048358-0700 CanBikeCO[74178:8688001] DEBUG: In checker's didChangeAuthorizationStatus, new authorization status = 3, always = 3
2022-07-05 22:47:07.050257-0700 CanBikeCO[74178:8688001] No foreground delegate found, calling SensorControlBackgroundChecker from didChangeAuthorizationStatus to verify location service status and permission
2022-07-05 22:47:07.050413-0700 CanBikeCO[74178:8688001] DEBUG: No foreground delegate found, calling SensorControlBackgroundChecker from didChangeAuthorizationStatus to verify location service status and permission
```

We call the background checker and because only the loc checks pass, we generate a redirect to the control screen.

```
2022-07-05 22:47:07.058763-0700 CanBikeCO[74178:8688001] Returning combination of always = 1 and precise 1
2022-07-05 22:47:07.058918-0700 CanBikeCO[74178:8688001] DEBUG: Returning combination of always = 1 and precise 1
2022-07-05 22:47:07.064496-0700 CanBikeCO[74178:8688001] DEBUG: Found existing field {
    redirectParams =     {
        launchAppStatusModal = 1;
    };
    redirectTo = "root.main.control";
}, retaining existing

```

**Comment by shankari** [comment]
there are potentially many errors here, but let's start with the first and most obvious. When the user selects "fix" and we launch the settings screen, we should not generate an error based on the initial auth status update to `0` because the user has not had a chance to fix the screen yet.

I assume that this is generated when we start `[[TripDiaryStateMachine instance].locMgr startUpdatingLocation];`
Actually not, I commented it out and I still get the callback, although I no longer see the location option in the app settings.

```
2022-07-05 23:06:49.770486-0700 CanBikeCO[79125:8715754] iOS 13+ detected, launching UI settings to easily enable always
2022-07-05 23:06:49.773236-0700 CanBikeCO[79125:8715754] THREAD WARNING: ['DataCollection'] took '27.124023' ms. Plugin should use a background thread.
2022-07-05 23:06:49.773787-0700 CanBikeCO[79125:8715754] In checker's didChangeAuthorizationStatus, new authorization status = 0, always = 3
2022-07-05 23:06:49.773907-0700 CanBikeCO[79125:8715754] DEBUG: In checker's didChangeAuthorizationStatus, new authorization status = 0, always = 3
```

Added additional logging and confirmed that we get the notification:
- irrespective of the `startUpdatingLocation` call (occurs even when it is commented out)
- after the app settings are launched 

```
2022-07-05 23:13:33.764785-0700 CanBikeCO[80555:8727042] iOS 13+ detected, launching UI settings to easily enable always
2022-07-05 23:13:33.764910-0700 CanBikeCO[80555:8727042] Opening app settings for the app using app-settings:
2022-07-05 23:13:33.768231-0700 CanBikeCO[80555:8727042] THREAD WARNING: ['DataCollection'] took '31.011230' ms. Plugin should use a background thread.
2022-07-05 23:13:33.769749-0700 CanBikeCO[80555:8727042] In checker's didChangeAuthorizationStatus, new authorization status = 0, always = 3
2022-07-05 23:13:33.770029-0700 CanBikeCO[80555:8727042] DEBUG: In checker's didChangeAuthorizationStatus, new authorization status = 0, always = 3
2022-07-05 23:13:33.772682-0700 CanBikeCO[80555:8727042] 1 foreground delegates found, calling didChangeAuthorizationStatus to return the new value 0
```

Although commenting out the app settings call didn't get rid of it either.

**Comment by shankari** [comment]
ok so we initialize the FSM again from `[self promptForPermission:[TripDiaryStateMachine instance].locMgr];` which seems redundant given that we just re-read the locMgr inside `promptForPermission`. It looks like added it to support the background call from the remote push. but the related code is commented out, so I am not sure it is actually needed.

Initializing the FSM doesn't generate the auth callback, though. And we didn't even launch the screen so that is not it.

```
2022-07-06 08:15:57.864065-0700 CanBikeCO[88292:8777046] DEBUG: initializing TripDiaryStateMachine with state = STATE_START
2022-07-06 08:16:42.024669-0700 CanBikeCO[88292:8777046] iOS 13+ detected, launching UI settings to easily enable always
2022-07-06 08:17:10.198555-0700 CanBikeCO[88292:8777046] THREAD WARNING: ['DataCollection'] took '472400.773193' ms. Plugin should use a background thread.
2022-07-06 08:17:50.106401-0700 CanBikeCO[88292:8777046] In checker's didChangeAuthorizationStatus, new authorization status = 0, always = 3
```

**Comment by shankari** [comment]
per https://stackoverflow.com/questions/7266700/when-does-an-iphone-application-receive-didchangeauthorizationstatus-delegate-c this can happen when we create a location manager or when we register the delegate.

**Comment by shankari** [comment]
ok, so it looks like there is really no alternative to getting that initial callback.
- It is generated whenever we register the delegate, 
- it is generated during the registration even if the value is unchanged, and
- we have to register the delegate when we launch the app so that we can get callbacks for out-of-band permission changes.

The delegate registration is not a problem when there is no foreground delegate, since we just launch the background checks. However, if there is a foreground delegate, then we "handle" it and then reset it, so the subsequent real change from the settings screen is missed.

We need to track whether the callback is from the delegate registration or from a user interaction.

Or... we could ensure that the delegate registration happens when we don't have the foreground delegate registered.
Let's start with reverting the change in https://github.com/e-mission/e-mission-docs/issues/735#issuecomment-1174732088

**Comment by shankari** [comment]
After reverting the change, we do get the initial callback when we launch the app, and before the foreground delegate is registered.

```
2022-07-09 22:52:51.525501-0700 CanBikeCO[68806:12276343] In checker's didChangeAuthorizationStatus, new authorization status = 0, always = 3
2022-07-09 22:52:51.525622-0700 CanBikeCO[68806:12276343] DEBUG: In checker's didChangeAuthorizationStatus, new authorization status = 0, always = 3
2022-07-09 22:52:51.527394-0700 CanBikeCO[68806:12276343] No foreground delegate found, calling SensorControlBackgroundChecker from didChangeAuthorizationStatus to verify location service status and permission
2022-07-09 22:52:51.527482-0700 CanBikeCO[68806:12276343] DEBUG: No foreground delegate found, calling SensorControlBackgroundChecker from didChangeAuthorizationStatus to verify location service status and permission
```

That works

```
2022-07-09 22:56:37.075917-0700 CanBikeCO[68806:12276343] In checker's didChangeAuthorizationStatus, new authorization status = 3, always = 3
2022-07-09 22:56:37.076065-0700 CanBikeCO[68806:12276343] DEBUG: In checker's didChangeAuthorizationStatus, new authorization status = 3, always = 3
2022-07-09 22:56:38.673219-0700 CanBikeCO[68806:12276343] 1 foreground delegates found, calling didChangeAuthorizationStatus to return the new value 3
2022-07-09 22:56:38.673427-0700 CanBikeCO[68806:12276343] DEBUG: 1 foreground delegates found, calling didChangeAuthorizationStatus to return the new value 3
2022-07-09 22:56:40.193365-0700 CanBikeCO[68806:12276343] In delegate didChangeAuthorizationStatus, newStatus = 3
2022-07-09 22:56:40.193495-0700 CanBikeCO[68806:12276343] DEBUG: In delegate didChangeAuthorizationStatus, newStatus = 3
2022-07-09 22:56:40.195398-0700 CanBikeCO[68806:12276343] status == always, restarting FSM if start state
2022-07-09 22:56:40.195527-0700 CanBikeCO[68806:12276343] DEBUG: status == always, restarting FSM if start state
2022-07-09 22:56:40.197239-0700 CanBikeCO[68806:12276343] Still in start state, sending initialize...
2022-07-09 22:56:40.197349-0700 CanBikeCO[68806:12276343] INFO: Still in start state, sending initialize...
2022-07-09 22:56:40.198736-0700 CanBikeCO[68806:12276343] Received platform-specific notification T_INITIALIZE
2022-07-09 22:56:40.198828-0700 CanBikeCO[68806:12276343] DEBUG: Received platform-specific notification T_INITIALIZE
2022-07-09 22:56:40.200082-0700 CanBikeCO[68806:12276343] In TripDiaryStateMachine, received transition T_INITIALIZE in state STATE_START
2022-07-09 22:56:40.200189-0700 CanBikeCO[68806:12276343] DEBUG: In TripDiaryStateMachine, received transition T_INITIALIZE in state STATE_START
2022-07-09 22:56:40.201615-0700 CanBikeCO[68806:12276343] data has 79 bytes, str has size 79
2022-07-09 22:56:40.202971-0700 CanBikeCO[68806:12276343] data has 69 bytes, str has size 69
2022-07-09 22:56:40.204495-0700 CanBikeCO[68806:12276343] registered for visit notifications
2022-07-09 22:56:40.204590-0700 CanBikeCO[68806:12276343] DEBUG: registered for visit notifications
2022-07-09 22:56:40.221607-0700 CanBikeCO[68806:12276343] In createGeofenceHere
2022-07-09 22:56:40.221725-0700 CanBikeCO[68806:12276343] DEBUG: In createGeofenceHere
2022-07-09 22:56:40.224672-0700 CanBikeCO[68806:12276343] currLoc = ... Pacific Daylight Time, which is current, can use it
2022-07-09 22:56:40.224762-0700 CanBikeCO[68806:12276343] DEBUG: currLoc = ... Pacific Daylight Time, which is current, can use it
2022-07-09 22:56:40.226171-0700 CanBikeCO[68806:12276343] data has 71 bytes, str has size 71
2022-07-09 22:56:40.227353-0700 CanBikeCO[68806:12276343] BEFORE creating region
2022-07-09 22:56:40.227439-0700 CanBikeCO[68806:12276343] DEBUG: BEFORE creating region
2022-07-09 22:56:40.232135-0700 CanBikeCO[68806:12276343] DEBUG: AFTER creating region
2022-07-09 22:56:42.407438-0700 CanBikeCO[68806:12276343] DEBUG: Notified all foreground delegates, removing all of them
```

So that works!

Only change was to add a `restartFSMIfStartState` even in the foreground delegate


```
            [LocalNotificationManager addNotification:[NSString stringWithFormat:@"status == always, restarting FSM if start state"]];
            [SensorControlBackgroundChecker restartFSMIfStartState];
```

**Comment by shankari** [comment]
Since we are mucking around with the start anyway, let's see if we can make a few more changes.
1. we should really not attempt to start reading the location until we get the consent, so changed the `oneTimeInitTracking` to only happen once we have consent.

```
    if ([ConfigManager getPriorConsent] != NULL) {
        [TripDiaryActions oneTimeInitTracking:CFCTransitionInitialize withLocationMgr:self.locMgr];
    }
```

This works. We first start in `STATE_START`

```
2022-07-10 08:39:05.131145-0700 emission[37411:12680294] in getPriorConsent, currConfig = (null)
2022-07-10 08:39:05.142625-0700 emission[37411:12680294] initializing TripDiaryStateMachine with state = STATE_START
```

we then start tracking so that the location turns up in the app setting screen, but get errors


```
2022-07-10 08:46:25.965490-0700 emission[37411:12680294] Location tracking failed with error Error Domain=kCLErrorDomain Code=1 "(null)"
2022-07-10 08:46:25.965658-0700 emission[37411:12680294] DEBUG: Location tracking failed with error Error Domain=kCLErrorDomain Code=1 "(null)"
```

we then get the new auth status and restart the FSM


```
2022-07-10 08:46:26.465572-0700 emission[37411:12680294] In checker's didChangeAuthorizationStatus, new authorization status = 3, always = 3
2022-07-10 08:46:26.467815-0700 emission[37411:12680294] 1 foreground delegates found, calling didChangeAuthorizationStatus to return the new value 3
2022-07-10 08:46:26.469602-0700 emission[37411:12680294] In delegate didChangeAuthorizationStatus, newStatus = 3
2022-07-10 08:46:26.472038-0700 emission[37411:12680294] status == always, restarting FSM if start state
```

and end up in the proper state

```
2022-07-10 08:46:27.075731-0700 emission[37411:12680294] Current state of region STATIONARY_GEOFENCE_LOCATION is 0 (unknown)
2022-07-10 08:46:27.077728-0700 emission[37411:12680294] Received UNKNOWN geofence state when currState = STATE_START, unclear what to do
2022-07-10 08:46:28.194639-0700 emission[37411:12680294] Application will enter the foreground
2022-07-10 08:46:29.113188-0700 emission[37411:12680294] Current state of region STATIONARY_GEOFENCE_LOCATION is 1 (inside)
2022-07-10 08:46:29.141281-0700 emission[37411:12680294] Moved from STATE_START to STATE_WAITING_FOR_TRIP_START
```

**Comment by shankari** [comment]
However, when the app first started, back when we had just launched, we still got a `GEOFENCE_CREATION_ERROR`

```
2022-07-10 08:39:05.175227-0700 emission[37411:12680294] Received platform-specific notification T_GEOFENCE_CREATION_ERROR
2022-07-10 08:39:05.213412-0700 emission[37411:12680294] In TripDiaryStateMachine, received transition T_GEOFENCE_CREATION_ERROR in state STATE_START
```

On adding breakpoints, we determined that this was from

```
+(void)checkAppState
{
...
   else {
        /*
         Log.i(ctxt, TAG, "location settings are valid, but location permission is not, generating tracking error and visible notification");
         Log.i(ctxt, TAG, "curr status check results = " +
                " loc permission, motion permission, notification, unused apps "+ Arrays.toString(allOtherChecks));
         */
        // Should replace with TRACKING_ERROR but looks like we
        // don't have any
        [[NSNotificationCenter defaultCenter]
            postNotificationName:CFCTransitionNotificationName
                object:CFCTransitionGeofenceCreationError];
        [self generateOpenAppSettingsNotification];
    }
}
```

which in turn is called when we get the initial `didChangeAuthorizationStatus` callback from the delegate registration.

I guess it is not bad to check the status before consent, but we should probably not set any notifications since the user has not consented. Changing the error prompts to also be only after consent.

**Comment by shankari** [comment]
After changing the app state checks to include

```
    else {
        if ([ConfigManager getPriorConsent] == NULL) {
            [LocalNotificationManager addNotification:[NSString stringWithFormat:@"User has not yet consented, ignoring failed checks"]];
        } else {
```

we get 

```

```
2022-07-10 09:42:31.304749-0700 emission[51132:12770695] In checker's didChangeAuthorizationStatus, new authorization status = 0, always = 3
2022-07-10 09:42:31.307270-0700 emission[51132:12770695] No foreground delegate found, calling SensorControlBackgroundChecker from didChangeAuthorizationStatus to verify location service status and permission
2022-07-10 09:42:31.310908-0700 emission[51132:12770695] Returning combination of always = 0 and precise 0
2022-07-10 09:42:31.317099-0700 emission[51132:12770695] in getPriorConsent, currConfig = (null)
2022-07-10 09:42:31.319301-0700 emission[51132:12770695] User has not yet consented, ignoring failed checks
```

Changing the location still works because we manually initialize

```
2022-07-10 09:45:25.579187-0700 emission[51132:12770695] Current state of region STATIONARY_GEOFENCE_LOCATION is 1 (inside)
2022-07-10 09:45:25.581934-0700 emission[51132:12770695] Received INSIDE geofence state when currState = STATE_START
2022-07-10 09:45:25.585108-0700 emission[51132:12770695] Received platform-specific notification T_INIT_COMPLETE
2022-07-10 09:45:25.587679-0700 emission[51132:12770695] In TripDiaryStateMachine, received transition T_INIT_COMPLETE in state STATE_START
2022-07-10 09:45:25.595650-0700 emission[51132:12770695] Moved from STATE_START to STATE_WAITING_FOR_TRIP_START
```

Note that there is still a minor gap where if the user changes the location to always and then changes it back to never *before consenting* we don't prompt anything. But in that case, the status screen is red, so I don't think that is a huge issue.

Also improved the logging around the check failures. So after consent, if the user changes the setting, we get the notification but also see logs

```
2022-07-10 09:49:29.072031-0700 emission[51132:12770695] allChecksPass = 0, but location checks fail, generating error and notification, allChecks: [loc settings, loc permissions, motion settings, motion permissions, notification] = (
    1,
    0,
    1,
    1,
    1
)
```

**Comment by shankari** [comment]
since we are mucking around here anyway, let's see if we can handle the commented out calls to the sensor control background checker.

```
    // STATUS SCREEN: Figure out how to fix recursion when [TripDiaryStateMachine instance] is called
    // [SensorControlBackgroundChecker restartFSMIfStartState];
```

The problem here is that the `SensorControlBackgroundChecker` also initializes the TripDiaryStateMachine instance
https://github.com/e-mission/e-mission-docs/issues/735#issuecomment-1174732088

```
+(BOOL)checkLocationPermissions {
    BOOL alwaysPerm = [CLLocationManager authorizationStatus] == kCLAuthorizationStatusAuthorizedAlways;
    BOOL precisePerm = TRUE;
            CLLocationManager* currLocMgr = [TripDiaryStateMachine instance].locMgr;
            if (@available(iOS 14.0, *)) {
                CLAccuracyAuthorization preciseOrNot = [currLocMgr accuracyAuthorization];
        precisePerm = preciseOrNot == CLAccuracyAuthorizationFullAccuracy;
            } else {
                [LocalNotificationManager addNotification:[NSString stringWithFormat:@"No precise location check needed for iOS < 14"]];
            }
    [LocalNotificationManager addNotification:[NSString stringWithFormat:@"Returning combination of always = %@ and precise %@", @(alwaysPerm), @(precisePerm)]];
    return alwaysPerm && precisePerm;
}
```

This recursive initialization crashes the system.

The obvious first try to avoid this initialization is to not initialize the state machine from the background checker. But we already know that this is not a good idea.
https://github.com/e-mission/e-mission-docs/issues/735#issuecomment-1179661743

The next try is to call the `SensorControlBackgroundChecker` in the places where the `TripDiaryStateMachine init` is called, but outside of the function.

These locations are:

```
./BEMDataCollection.m:    self.tripDiaryStateMachine = [TripDiaryStateMachine instance];
./Verification/SensorControlForegroundDelegate.m:            [self promptForPermission:[TripDiaryStateMachine instance].locMgr];
./Verification/SensorControlForegroundDelegate.m:    [[TripDiaryStateMachine instance].locMgr stopUpdatingLocation];
./Verification/SensorControlForegroundDelegate.m:        [[TripDiaryStateMachine instance].locMgr startUpdatingLocation];
./Verification/SensorControlBackgroundChecker.m:    NSUInteger currState = [TripDiaryStateMachine instance].currState;
./Verification/TripDiarySensorControlChecks.m:            CLLocationManager* currLocMgr = [TripDiaryStateMachine instance].locMgr;
```

Of these, the
- `SensorControlForegroundDelegate` should not affect background checks since they are intended to be launched by the user
- `TripDiarySensorControlChecks` is a generic callback that can be invoked from anywhere

- `BEMDataCollection` is the only location which is invoked directly and periodically, and we may want to add a new background check 


```
- (void)initWithConsent {
    self.tripDiaryStateMachine = [TripDiaryStateMachine instance];
    [SensorControlBackgroundChecker checkAppState];
    NSDictionary* emptyOptions = @{};
    [AppDelegate didFinishLaunchingWithOptions:emptyOptions];
}
```

works

```
2022-07-10 11:10:39.167961-0700 emission[69695:12885598] Moved from STATE_START to STATE_WAITING_FOR_TRIP_START
```

**Comment by shankari** [comment]
@asiripanich I have fixed this now if you want to pull and rebuild.
